### ISSUE TITLE
PROPOSAL: Pull payments with TSP and Light Connectors

### ISSUE THREAD

**Issue by sharafian** [interledger/rfcs#156]
(pasted from gist, to open up discussion better)

# Pull Payments with Teaspoon and Light Connectors

> A monk asked Evan: 'Without pulling, without revoking, how can you express the payment?'
> - _The Gateless Gateway_

# Flow

Like a lot of things in Interledger, pull payments can be though of as circular Interledger payments.
A customer (C) extends some amount of trust to a merchant (M).
M uses this trust in order to get credit on some ledger, which C may or may not be on.

We already have a convenient way to represent trust in our network: trustlines.
If C and M have a trustline on which M has a positive balance, we can say that C has extended trust to M.

Right now, trustlines are created in the context of peering relationships.
In the pull-payments use-case, M must request that C creates a trustline for a certain amount.
Additionally, the trustline may have to replenish at some rate (e.g. a subscription).

How do we set up this trustline with M's desired properties, and with a good experience for C?

## Trustline Setup Protocol

The Trustline Setup Protocol (abbreviated TSP or Teaspoon), allows a merchant to request a trustline from a customer.
TSP is initiated by the merchant, and goes through a flow that looks a lot like a current online checkout.

- The user gives the merchant their ILP identifier (e.g. `alice@example.com`)
- The merchant makes a `POST` to `https://example.com/.well-known/ilp/tsp?account=alice`, with data:

```js
{
  // quote this amount in order to calculate trustline balance (TODO: what if rates change?)
  "allowance_destination_amount": "10.00",
  "allowance_destination_account": "g.crypto.bitcoin.hcntq4t68xdg45hu89dj78cjj4t3hr",
  // information on how the customer should manage their trustline
  "frequency": 'P1M', // ISO 8601 duration (https://en.wikipedia.org/wiki/ISO_8601#Durations)
  "minimum_destination_amount": "10.00", // indicates that the whole allowance will be spent at once
  "maximum_destination_amount": "10.00",
  "public_key": "AIMOAGWNFUYIBYiopjuNO78QSFtXgOwP1sbsCSk", // for the trustline's shared secret
  "metadata": {
    // merchant information to display, as receiver info in SPSP
  },
  "webhooks": {
    // webhooks for `example.com` to call back to after the customer reviews
    "approved": "https://merchant.com/payment/252643636/approved",
    "rejected": "https://merchant.com/payment/252643636/rejected"
  }
}
```

- `example.com` responds to this request with:

```js
{
  "redirect": "https://example.com/api/tsp/approval?id=15230572626263"
}
```

- Merchant redirects the customer to `"redirect"`.
- The customer reviews the details on `example.com` and decides whether to approve or reject the request.
- If the customer rejects the request, then the customer is sent to the `"rejected"` webhook, and TSP is over.
- If the customer approves the request, then the customer is sent to the `"approved"` webhook, which is called with:

```js
{
  "prefix": "peer.AWhD7.usd.",
  "public_key": "KRixgcBCBdyQln7IBYiopjuNO78QSFtXgOwP1sbsCSk",
  "rpc_uri": "https://example.com/api/receivers/alice/rpc"
}
```

The trustline is now complete, and ready for the merchant to use.

## How the Merchant Uses the Trustline

When making a "pull payment," M uses their trustline with C to send an interledger payment to themself.
This decreases the trustline balance, and causes C to pay out on another ledger (we'll get to how that works
in a bit).

M is simultaneously sender and receiver, which means they can use any transport protocol that they want,
without worrying about how the information is transmitted. For example, M could "pull" 10 dollars like so:

```js
const receiver = ILP.createReceiver({ /* ... */ })
const sender = ILP.createSender({
  _plugin: 'ilp-plugin-virtual',
  // trustline details from TSP
})

// passes request in memory instead of over HTTP
const request = receiver.createRequest({
  destination_amount: '10.00'
})
const quote = sender.quoteRequest(request)

receiver.listen()
sender.payRequest(quote)
receiver.on('incoming', (transfer) => {
  console.log('ayy received transfer')
  // process payment and clean up sender and receiver
  // ...
})
```

## Light Connectors: How the Customer's ILP Kit routes the payment

In order to receive a payment from a trustline and pay out on another ledger, the receiver needs a connector of some kind.
However, running the full `ilp-connector` would be overkill.
The receiver doesn't need cross-currency raates, nor do they need a routing table.
Everything that comes in on the trustline gets paid out on the connector's other plugin.
The end result is that this looks more like `ilp-plugin-settlement-adapter` than `ilp-connector`.
We can create a component for this purpose, called `ilp-connector-light`.

### Behavior of `ilp-connector-light`

Assume the light connector has 2 plugins, A and B.

- On an incoming payment from A, the light connector quotes the `ilp_header` on B, and sends the resulting payment.
- On an incoming quote request on A, the light connector forwards the quote request to B, and returns the result on A.

Further features are related to trustline management. These could be handled by another stateful component (to track
long durations), or the light connector could be kept running and track things like subscriptions and allowances
internally, as part of its functionality.

# In Summary

With a protocol to request a trustline (TSP), a more minimal implementation of the connector (`ilp-connector-light`),
and circular ILP payments, we can implement pull payments without otherwise changing our components.

**Comment by adrianhopebailie** [comment]
> we can implement pull payments without otherwise changing our components.

Is this a good goal? I'd consider re-use of components a low priority. On that note, do we have a clear set of goals/requirements? Why is a such a mechanism needed?

@sharafian this is a clever solution but seems very complicated when all you really need is the ability to delegate permission to a third party to execute payments off your ledger. Think OAuth but with some extra meta data (like amount, frequency, etc) tied to the auth token the RP gets.

The challenge with ILP-based pull payments is mostly related to changing rates between the sending and receiving ledgers as opposed to the delegation of trust (which is many respects is a solved problem, or being solved, independent of ILP).

**Comment by sharafian** [comment]
If the OAuth-like system you're proposing authenticates to the customer's ledger, then that would make our ledger API a lot more complicated. Additionally, the merchant would need ledger plugins for every ledger that might want to send to them (in order to authenticate to their ledgers)

On the other hand, if the OAuth-like trust system is its own component, it would look a lot like what I'm proposing. It would have some API for getting a token (TSP), some system for tracking a token's balance (`plugin-virtual`), and would turn API calls into outgoing payment's on the customer's ledger (`plugin-virtual`'s RPC + light connector function exactly this way)

I agree that the changing rates is a big problem either way, though, and something we need to address.

**Comment by emschwartz** [comment]
To clarify the motivation for this: Having a way to do pull payments would dramatically improve the experience in a number of important use cases. The main one is for 3rd party apps that want to integrate ILP payments. Right now you need the user to put their ledger account credentials into the app, and you need a ledger plugin that supports their ledger. Both of these make developing apps with ILP nearly impossible. 👎  Pull payments would also support subscription and recurring payment use cases.

Pull payments should definitely be implemented as a separate protocol and service so that you could run a pull payment server for yourself even if the ledger does not support pull payments (e.g. bitcoin). The question is whether we should use OAuth 1 (used by the [Open Bank Project](https://github.com/OpenBankProject/OBP-API/wiki#direct-login--jwt) and [MasterCard](https://mastercarddevelopersupport.desk.com/customer/portal/articles/search?q=authentication+requirements)), OAuth 2 (used by [PayPal](https://developer.paypal.com/docs/api/auth-headers/) and [Coinbase](https://developers.coinbase.com/docs/wallet/coinbase-connect/permissions)) or TSP.

Right now I'm leaning towards being in favor of something like TSP, mostly because as @sharafian said we'll end up with similar but different components if we try to implement this all using OAuth. We need:
- **API endpoint discovery** (either resolving a `user@host` address supplied by the user, or saved by the browser if possible) - not handled by OAuth, because in most OAuth use cases you know what identity provider to direct the user to
- **Scopes / limits** - not exactly handled by OAuth. OAuth has scopes but we'd need to specify how you request a specific spending limit or the frequency with which the limit is expressed. Note the only OAuth API I could find that includes anything like a spending limit is [Coinbase](https://developers.coinbase.com/docs/wallet/coinbase-connect/permissions).
- **User authorization flow** - probably want something similar to what OAuth has whether we use that protocol or not
- **API for transfers and messages** - nothing to do with OAuth. `ilp-plugin-virtual` uses an API similar to the ledger plugin interface, which is already meant to be an abstraction over different types of underlying ledgers. We'd need something similar in this case.
- **API request authentication** - OAuth 1 uses [signatures](https://oauth.net/core/1.0a/#signing_process) while OAuth 2 uses HTTP [bearer tokens](http://self-issued.info/docs/draft-ietf-oauth-v2-bearer.html#anchor5)

**Comment by adrianhopebailie** [comment]
> Pull payments should definitely be implemented as a separate protocol and service so that you could run a pull payment server for yourself even if the ledger does not support pull payments (e.g. bitcoin).

Good requirement. +1

> The question is whether we should use OAuth 1 (used by the Open Bank Project and MasterCard), OAuth 2 (used by PayPal and Coinbase) or TSP.

Worded as a requirement: _We need a protocol for delegating authority to initiate transfer to a third-party._ 

Some protocols for this exist already but we may have requirements they don't meet that are so difficult to implement on top of the existing protocols that it's worth just inventing our own (sounds like a pretty big assumption).

> with similar but different components if we try to implement this all using OAuth

Re-using some existing components for a "a separate protocol and service" seems like a very bad motivation for a technology choice. Not to mention that the libraries and components already out in the ecosystem that support existing delegation protocols are VERY mature and battle hardened. Pretty much every major online service uses some form of OAuth for delegation. Build-your-own security is seldom a good idea.

> (either resolving a user@host address supplied by the user, or saved by the browser if possible) - not handled by OAuth, because in most OAuth use cases you know what identity provider to direct the user to

Not true. WebFinger was developed exactly for the use case of discovering an IdP endpoint from an `acct:` URI or `user@host` identifier. See [OpenID Connect Issuer Discovery](https://openid.net/specs/openid-connect-discovery-1_0.html#IssuerDiscovery).

Before we dive into picking a solution can we define the requirement. Is the requirement that we start with an `account@host` identifier or can we start with a URL? If we're determined to design our own WebFinger protocol at `.well-known/spsp` why not just make it `.well-known/ilp` and let that end point return everything about an account that is required to send to and from it (i.e. sender OAuth/TSP endpoint and SPSP receiver endpoint).

> Scopes / limits - not exactly handled by OAuth.

No, but as you say it's been done by Coinbase and is being used on a production system so we shouldn't discard it immediately. We're stepping into territory that many have already been working in for a while, we should at least try to use some existing standards before we design this ourselves.

> API for transfers and messages - nothing to do with OAuth. ilp-plugin-virtual uses an API similar to the ledger plugin interface, which is already meant to be an abstraction over different types of underlying ledgers. We'd need something similar in this case.

We should distinguish between getting delegated auth to transact and transacting. It's quite likely that the transaction API will be exactly the same whether the system accessing the account is doing it with the account owner's credentials or delegated credentials.

**Comment by emschwartz** [comment]
> Is the requirement that we start with an account@host identifier or can we start with a URL?

We could start with a URL. We need to think about how the merchant would get that from the user though. If a browser could tell the merchant, then a URL would be fine, but then we depend on browser support or extensions. If it's a `user@host` identifier, the user could put that in.

> why not just make it .well-known/ilp and let that end point return everything about an account that is required to send to and from it

We've talked about this and I think it's not a bad idea.

> It's quite likely that the transaction API will be exactly the same whether the system accessing the account is doing it with the account owner's credentials or delegated credentials.

This will only be true if we say the pull payment API expects the Common Ledger API and just uses OAuth with that. Not a bad idea... :thinking:

**Comment by michielbdejong** [comment]
I used to be a big fan of OAuth, but I'm not anymore. It's very web-centric, meaning it's just not very relevant when you think about software outside the browser. I consider myself a "web guy", but I have to admit that in 2017, a lot of software seems to be running on servers and as native/installed applications on (mobile) devices, so the "redirect dance" that OAuth gives you is just not as relevant these days as it seemed to be five years ago.

If I allow a shop to do pull payments from my account then I would like to set the following requirements:
* the shop is not blacklisted by my ledger
* the amount pulled per transaction does not exceed X
* the number of individual transactions does not exceed X per day/week/month
* the total amount pulled does not exceed X per day/week/month
* each transaction has a clear receipt, specifying what it's for
* I can stop the authorization at any time, so that no further pull payments will occur
* I can request a refund from the shop (ideally, through a simple click in the UI of my ledger)
* I can phone/visit/get-angry-at/twitter-bomb/sue the shop if they ignore my refund request

Being able to set up the pull payment using a web browser (redirect between ilp-kit and webshop) is not a central requirement, I would say? And if it is a requirement, then shouldn't we try to do this within PaymentRequest instead?

**Comment by adrianhopebailie** [comment]
> I used to be a big fan of OAuth, but I'm not anymore. It's very web-centric, meaning it's just not very relevant when you think about software outside the browser. I consider myself a "web guy", but I have to admit that in 2017, a lot of software seems to be running on servers and as native/installed applications on (mobile) devices, so the "redirect dance" that OAuth gives you is just not as relevant these days as it seemed to be five years ago.

This is true but there are all sorts of security issues you have to deal with when you are doing delegated authority if you have no control over the interface your user sees when agreeing to the delegation.

OAuth relies on the fact that the user is redirected to a website hosted by the IdP so the RP relinquishes control of the session temporarily. There are other ways to get the user's consent that are secure (like one-time-PINs delivered by another secure channel) but we'd need to solve for that then.
